File: win-prog/ProgWinVB/Edit List and Spin/NotepadCloneWithFormat/NotepadCloneWithFormat.vb
'-------------------------------------------------------
' NotepadCloneWithFormat.vb (c) 2002 by Charles Petzold
'-------------------------------------------------------
Imports Microsoft.Win32
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class NotepadCloneWithFormat
    Inherits NotepadCloneWithEdit

    ' Strings for registry
    Const strWordWrap As String = "WordWrap"
    Const strFontFace As String = "FontFace"
    Const strFontSize As String = "FontSize"
    Const strFontStyle As String = "FontStyle"
    Const strForeColor As String = "ForeColor"
    Const strBackColor As String = "BackColor"
    Const strCustomClr As String = "CustomColor"

    Private clrdlg As New ColorDialog()
    Private miFormatWrap As MenuItem

    Shared Shadows Sub Main()
        Application.Run(New NotepadCloneWithFormat())
    End Sub

    Sub New()
        strProgName = "Notepad Clone with Format"
        MakeCaption()

        ' Format
        Dim mi As New MenuItem("F&ormat")
        AddHandler mi.Popup, AddressOf MenuFormatOnPopup
        Menu.MenuItems.Add(mi)
        Dim index As Integer = Menu.MenuItems.Count - 1

        ' Format Word Wrap
        miFormatWrap = New MenuItem("&Word Wrap")
        AddHandler miFormatWrap.Click, AddressOf MenuFormatWrapOnClick
        Menu.MenuItems(index).MenuItems.Add(miFormatWrap)

        ' Format Font
        mi = New MenuItem("&Font...")
        AddHandler mi.Click, AddressOf MenuFormatFontOnClick
        Menu.MenuItems(index).MenuItems.Add(mi)

        ' Format Background Color
        mi = New MenuItem("Background &Color...")
        AddHandler mi.Click, AddressOf MenuFormatColorOnClick
        Menu.MenuItems(index).MenuItems.Add(mi)
    End Sub

    Protected Overrides Sub OnLoad(ByVal ea As EventArgs)
        MyBase.OnLoad(ea)

        ' Help
        Dim mi As New MenuItem("&Help")
        Menu.MenuItems.Add(mi)
        Dim index As Integer = Menu.MenuItems.Count - 1

        ' Help About
        mi = New MenuItem("&About " & strProgName & "...")
        AddHandler mi.Click, AddressOf MenuHelpAboutOnClick
        Menu.MenuItems(index).MenuItems.Add(mi)
    End Sub

    Private Sub MenuFormatOnPopup(ByVal obj As Object, ByVal ea As EventArgs)
        miFormatWrap.Checked = txtbox.WordWrap
    End Sub

    Private Sub MenuFormatWrapOnClick(ByVal obj As Object, _
                                      ByVal ea As EventArgs)
        Dim mi As MenuItem = DirectCast(obj, MenuItem)
        mi.Checked = Not mi.Checked
        txtbox.WordWrap = mi.Checked
    End Sub

    Private Sub MenuFormatFontOnClick(ByVal obj As Object, _
                                      ByVal ea As EventArgs)
        Dim fntdlg As New FontDialog()
        fntdlg.ShowColor = True
        fntdlg.Font = txtbox.Font
        fntdlg.Color = txtbox.ForeColor
        If fntdlg.ShowDialog() = DialogResult.OK Then
            txtbox.Font = fntdlg.Font
            txtbox.ForeColor = fntdlg.Color
        End If
    End Sub

    Private Sub MenuFormatColorOnClick(ByVal obj As Object, _
                                       ByVal ea As EventArgs)
        clrdlg.Color = txtbox.BackColor
        If clrdlg.ShowDialog() = DialogResult.OK Then
            txtbox.BackColor = clrdlg.Color
        End If
    End Sub

    Private Sub MenuHelpAboutOnClick(ByVal obj As Object, _
                                     ByVal ea As EventArgs)
        MessageBox.Show(strProgName & " " & Chr(169) & _
                        " 2002 by Charles Petzold", strProgName)
    End Sub

    Protected Overrides Sub LoadRegistryInfo(ByVal regkey As RegistryKey)
        MyBase.LoadRegistryInfo(regkey)
        txtbox.WordWrap = CBool(regkey.GetValue(strWordWrap))
        txtbox.Font = New Font( _
            DirectCast(regkey.GetValue(strFontFace), String), _
            Single.Parse(DirectCast(regkey.GetValue(strFontSize), String)), _
            CType(regkey.GetValue(strFontStyle), FontStyle))
        txtbox.ForeColor = Color.FromArgb( _
                                    CInt(regkey.GetValue(strForeColor)))
        txtbox.BackColor = Color.FromArgb( _
                                    CInt(regkey.GetValue(strBackColor)))

        Dim i, aiColors(16) As Integer

        For i = 0 To 15
            aiColors(i) = CInt(regkey.GetValue(strCustomClr & i))
        Next i
        clrdlg.CustomColors = aiColors
    End Sub

    Protected Overrides Sub SaveRegistryInfo(ByVal regkey As RegistryKey)
        MyBase.SaveRegistryInfo(regkey)
        regkey.SetValue(strWordWrap, CInt(txtbox.WordWrap))
        regkey.SetValue(strFontFace, txtbox.Font.Name)
        regkey.SetValue(strFontSize, txtbox.Font.SizeInPoints.ToString())
        regkey.SetValue(strFontStyle, CInt(txtbox.Font.Style))
        regkey.SetValue(strForeColor, txtbox.ForeColor.ToArgb())
        regkey.SetValue(strBackColor, txtbox.BackColor.ToArgb())

        Dim i As Integer
        For i = 0 To 15
            regkey.SetValue(strCustomClr & i, clrdlg.CustomColors(i))
        Next i
    End Sub
End Class

File: win-prog/ProgWinCS/Edit List and Spin/NotepadCloneWithFormat/NotepadCloneWithFormat.cs
//-----------------------------------------------------
// NotepadCloneWithFormat.cs © 2001 by Charles Petzold
//-----------------------------------------------------
using Microsoft.Win32;
using System;
using System.Drawing;
using System.Windows.Forms;

class NotepadCloneWithFormat: NotepadCloneWithEdit
{
	// Strings for registry

	const string strWordWrap  = "WordWrap";
	const string strFontFace  = "FontFace";
	const string strFontSize  = "FontSize";
	const string strFontStyle = "FontStyle";
	const string strForeColor = "ForeColor";
	const string strBackColor = "BackColor";
	const string strCustomClr = "CustomColor";

	ColorDialog clrdlg = new ColorDialog();
	MenuItem    miFormatWrap;

	public new static void Main()
	{
		Application.Run(new NotepadCloneWithFormat());
	}
	public NotepadCloneWithFormat()
	{
		strProgName = "Notepad Clone with Format";
		MakeCaption();

		// Format

		MenuItem mi = new MenuItem("F&ormat");
		mi.Popup += new EventHandler(MenuFormatOnPopup);
		Menu.MenuItems.Add(mi);
		int index = Menu.MenuItems.Count - 1;

		// Format Word Wrap

		miFormatWrap = new MenuItem("&Word Wrap");
		miFormatWrap.Click += new EventHandler(MenuFormatWrapOnClick);
		Menu.MenuItems[index].MenuItems.Add(miFormatWrap);

		// Format Font

		mi = new MenuItem("&Font...");
		mi.Click += new EventHandler(MenuFormatFontOnClick);
		Menu.MenuItems[index].MenuItems.Add(mi);

		// Format Background Color

		mi = new MenuItem("Background &Color...");
		mi.Click += new EventHandler(MenuFormatColorOnClick);
		Menu.MenuItems[index].MenuItems.Add(mi);
	}
	protected override void OnLoad(EventArgs ea)
	{
		base.OnLoad(ea);

		// Help

		MenuItem mi = new MenuItem("&Help");
		Menu.MenuItems.Add(mi);
		int index = Menu.MenuItems.Count - 1;

		// Help About

		mi = new MenuItem("&About " + strProgName + "...");
		mi.Click += new EventHandler(MenuHelpAboutOnClick);
		Menu.MenuItems[index].MenuItems.Add(mi);
	}
	void MenuFormatOnPopup(object obj, EventArgs ea)
	{
		miFormatWrap.Checked = txtbox.WordWrap;
	}
	void MenuFormatWrapOnClick(object obj, EventArgs ea)
	{
		MenuItem mi = (MenuItem) obj;
		mi.Checked ^= true;
		txtbox.WordWrap = mi.Checked;
	}
	void MenuFormatFontOnClick(object obj, EventArgs ea)
	{
		FontDialog fontdlg = new FontDialog();

		fontdlg.ShowColor = true;
		fontdlg.Font = txtbox.Font;
		fontdlg.Color = txtbox.ForeColor;

		if (fontdlg.ShowDialog() == DialogResult.OK)
		{
			txtbox.Font = fontdlg.Font;
			txtbox.ForeColor = fontdlg.Color;
		}
	}
	void MenuFormatColorOnClick(object obj, EventArgs ea)
	{
		clrdlg.Color = txtbox.BackColor;

		if (clrdlg.ShowDialog() == DialogResult.OK)
			txtbox.BackColor = clrdlg.Color;
	}
	void MenuHelpAboutOnClick(object obj, EventArgs ea)
	{
		MessageBox.Show(strProgName + " © 2001 by Charles Petzold", 
			strProgName);
	}
	protected override void LoadRegistryInfo(RegistryKey regkey)
	{
		base.LoadRegistryInfo(regkey);

		txtbox.WordWrap = Convert.ToBoolean(
			(int) regkey.GetValue(strWordWrap));
		txtbox.Font = new Font((string) regkey.GetValue(strFontFace),
			float.Parse(
			(string) regkey.GetValue(strFontSize)),
			(FontStyle) regkey.GetValue(strFontStyle));
		txtbox.ForeColor = Color.FromArgb(
			(int) regkey.GetValue(strForeColor));
		txtbox.BackColor = Color.FromArgb(
			(int) regkey.GetValue(strBackColor));
          
		int[] aiColors = new int[16];

		for (int i = 0; i < 16; i++)
			aiColors[i] = (int) regkey.GetValue(strCustomClr + i);

		clrdlg.CustomColors = aiColors;
	}
	protected override void SaveRegistryInfo(RegistryKey regkey)
	{
		base.SaveRegistryInfo(regkey);

		regkey.SetValue(strWordWrap,  Convert.ToInt32(txtbox.WordWrap));
		regkey.SetValue(strFontFace,  txtbox.Font.Name);
		regkey.SetValue(strFontSize,  txtbox.Font.SizeInPoints.ToString());
		regkey.SetValue(strFontStyle, (int) txtbox.Font.Style);
		regkey.SetValue(strForeColor, txtbox.ForeColor.ToArgb());
		regkey.SetValue(strBackColor, txtbox.BackColor.ToArgb());

		for (int i = 0; i < 16; i++)
			regkey.SetValue(strCustomClr + i, clrdlg.CustomColors[i]);
	}
}          
